### PR TITLE
Add serialization for Gen Dynamic DSL traces and some combinators

### PR DESCRIPTION
**Pull request by sritchie**

### CODE REVIEW

**Review by sritchie — COMMENTED** [review]
Looks like there are some empty files committed?

File: src/GenSerialization.jl [hunk 1]
```diff
@@ -1,5 +1,35 @@
 module GenSerialization
+using Gen
 
-# Write your package code here.
+"""
+Serialization produces a .gen file that relies on Julia's native serialization.
```
[the remark this answers is not in the document]
Comment (limarta) [reply src/GenSerialization.jl:5]: No we will use Julia's `Serialization` for now. It is the most stable and it's unlikely we will support cross-serialization anytime soon.

File: src/GenSerialization.jl [hunk 2]
```diff
@@ -1,5 +1,32 @@
 module GenSerialization
```
Comment (limarta) [inline src/GenSerialization.jl:1]: The latest commit has all of the requisites: (a) README has demo, installation, and warnings. (b) Serialization has been stripped down to the naive one mentioned long ago. An experimental speed ups have been removed.

Once this PR is merged, we should update Gen.jl to have this as a package extension.

**Review by limarta — COMMENTED** [review]
This PR contains all of the requisites: (a) A README that contains installation, usage, and warning instructions. (b) Removed experimental serialization code.

Once this PR is merged, we should add this as a package extension in Gen.jl

File: test/map.jl [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+@testset "Map" begin
+    tr = simulate(mapped_model, ([2,3,4],))
+    serialize("test.gen", tr)
+
+    recovered_tr = realize("test.gen", model)
+    # @test test_equality(tr, recovered_tr)
+
+    # Deserialization
+    # recovered_tr = deserialize("test.gen")
```
[the remark this answers is not in the document]
Comment (limarta) [reply test/map.jl]: Removed.

File: test/runtests.jl [hunk 4]
```diff
@@ -1,6 +1,19 @@
 using GenSerialization
+using Gen
 using Test
 
-@testset "GenSerialization.jl" begin
-    # Write your tests here.
+function test_equality(expected::Gen.DynamicDSLTrace, given)
+    """
+    TODO: Not an extensive set of tests.
```
[the remark this answers is not in the document]
Comment (limarta) [reply test/runtests.jl]: Removed.

File: src/serializers/dynamic/realization.jl [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    else
+        @warn "LOST KEY" key state.ptr_trie _module=""
+        throw("$(key) Key not in leaf or internal maps")
+    end
+
+
+    retval = record.subtrace_or_retval
+    # Check if it is truly a retval
+
+    # constrained = has_value(state.constraints, key)
```
[the remark this answers is not in the document]
Comment (limarta) [reply src/serializers/dynamic/realization.jl]: Removed.

File: src/serializers/dynamic/realization.jl [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    isempty = read(io, Bool)
+    score = read(io, Float64)
+    noise = read(io, Float64)
+    args = Serialization.deserialize(io)
+    retval = Serialization.deserialize(io)
+
+    @debug "DESERIALIZE" type=trace_type isempty score noise args retval gen_fn _module=""
+    ptr_trie = Gen.Trie{Any, RECORD_INFO}()
+    read_address_maps(io, ptr_trie, ())
+    if isempty
```
[the remark this answers is not in the document]
Comment (limarta) [reply src/serializers/dynamic/realization.jl]: Removed realization.

File: src/serializers/combinators/switch/realization.jl [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# mutable struct SwitchDeserializeState{T}
+#     score::Float64
+#     noise::Float64
+#     index::Int
+#     subtrace::Trace
+#     retval::T
+#     SwitchDeserializeState{T}(score::Float64, noise::Float64) where T = new{T}(score, noise)
+# end
+
+# function realize_trace(io::IO, gen_fn::Switch{C, N, K, T}) where {C, N, K, T}
```
[the remark this answers is not in the document]
Comment (limarta) [reply src/serializers/combinators/switch/realization.jl]: Re-added the combinator serializers. Should now be updated.

File: src/GenSerialization.jl [hunk 8]
```diff
@@ -1,5 +1,35 @@
...
+
+include("gen_file.jl")
+include("write_session.jl")
+include("serializers/serializers.jl")
+include("lazy_structs/lazy_structs.jl")
+include("file_header.jl")
+include("serialization.jl")
+include("deserialization.jl")
+
+# Native caching
```
[the remark this answers is not in the document]
Comment (limarta) [reply src/GenSerialization.jl:21]: Are you referring to the #native caching comment @sritchie ?

File: README.md [hunk 9]
```diff
@@ -1,3 +1,24 @@
...
+```
+@gen function model(p) 
+    x ~ bernoulli(p)
+end
+
+trace = simulate(model, (0.2))
+serialize("coin_flip.gen", trace)
+```
+
+2. To deserialize, there are currently two options. They different in performance, but the faster one currently requires non-DRY data structures.
```
[the remark this answers is not in the document]
Comment (limarta) [reply README.md]: Removed any mention of `realize`.

File: README.md [hunk 10]
```diff
@@ -1,3 +1,24 @@
-# GenSerialization
+# GenSerialization.jl
 
-[![Build Status](https://github.com/limarta/GenSerialization.jl/actions/workflows/CI.yml/badge.svg?branch=main)](https://github.com/limarta/GenSerialization.jl/actions/workflows/CI.yml?query=branch%3Amain)
+WIP tool to serialize Gen.jl traces using Julia's `serialization` library. Gen.jl traces often contain ephemeral data related to the generative model (a function), and this script decouples their dependency. Both traced and untraced randomness are considered.
+
+
+See the example for a typical use case. The general workflow is as follows:
+1. Produce traces from a dynamic DSL trace, and call the `serialize` function.
+```
```
[the remark this answers is not in the document]
Comment (limarta) [reply README.md]: In the README under installation.

File: src/GenSerialization.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (sritchie) [reply src/GenSerialization.jl:21]: yeah, correct. if it's clear for Julia people, sounds good to me.

[the remark this answers is not in the document]
Comment (limarta) [reply src/GenSerialization.jl:21]: Yes. This enables faster load times. https://discourse.julialang.org/t/precompile-why/78770

File: src/experimental/msgpack.jl [hunk 11]
```diff
@@ -0,0 +1,77 @@
+using MsgPack
+
+
+# class _MsgpackExtType(enum.IntEnum):
```
[the remark this answers is not in the document]
Comment (limarta) [reply src/experimental/msgpack.jl]: deleted

**Pull request merged**